File: Python/Recursion-2/split_array.py

# given an array of ints, is it possible to divide the ints into two groups, so that the sums of the two groups are the same
# every int must be in one group or the other
# write a recursive helper method that takes whatever arguments you like, and make the initial call to your recursive helper from split_array()
def split_array(nums):
    return split_array_helper(0, nums, 0, 0)

def split_array_helper(start, nums, group1, group2):
    if start >= len(nums):
        return group1 == group2
    else:
        return split_array_helper(start + 1, nums, group1 + nums[start], group2) or split_array_helper(start + 1, nums, group1, group2 + nums[start])

print(split_array([2, 2]))
print(split_array([2, 3]))
print(split_array([5, 2, 3]))

File: Java/Recursion-2/splitArray.java

public class recursion2 {
	
	// given an array of ints, is it possible to divide the ints into two groups, so that the sums of the two groups are the same
	// every int must be in one group or the other
	// write a recursive helper method that takes whatever arguments you like, and make the initial call to your recursive helper from splitArray()
	public boolean splitArray(int[] nums) {
		
		return splitArrayHelper(0, nums, 0, 0);
		
	}
	
	public boolean splitArrayHelper(int start, int[] nums, int group1, int group2) {
		
		if (start >= nums.length) {
			return group1 == group2;
		}
		else {
			return splitArrayHelper(start + 1, nums, group1 + nums[start], group2) || splitArrayHelper(start + 1, nums, group1, group2 + nums[start]);
		}
		
	}
	
	public static void main(String[] args) {
		
		recursion2 o = new recursion2();
		
		int[] nums1 = {2, 2};
		int[] nums2 = {2, 3};
		int[] nums3 = {5, 2, 3};
		
		System.out.println(o.splitArray(nums1));
		System.out.println(o.splitArray(nums2));
		System.out.println(o.splitArray(nums3));

	}

}
